### PR TITLE
Major refactoring of ISD-AS handling.

### PR DESCRIPTION
**Pull request by kormat**
- Rename AD to AS.
- Re-write ISD_AD as ISD_AS, and use everywhere.
- Remove a lot of junk docstrings, and trim others.
- Change a lot of code to "pragma: no cover", and remove the unit tests.
- Cleanup and standardize a lot of **str** representations.
- Improve the interface to path_db, simplifying lots of code that uses it.
- Change SCIONAddrHdr to use SCIONAddr
- Simplify InterfaceElement a bit, and improve naming.
- Merged CertChainRequest and TRCRequest, as they were ~identical.
- Removed the simulator sphinx indexes, as it breaks the sphinx build.

<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/630%23issuecomment-184278734%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r52927755%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23issuecomment-184325619%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r52982571%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23issuecomment-184665022%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53016669%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53018578%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23issuecomment-184735615%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53041776%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53041867%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23issuecomment-184278734%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40pszalach%20%3A%29%5Cr%5Cn%5Cr%5Cn%40aznair%20%40ercanucan%20-%20can%20you%20please%20review%20the%20changes%20in%20endhost/%2C%20please%3F%20I%20haven%27t%20tested%20anything%20other%20than%20sciond%20in%20there.%22%2C%20%22created_at%22%3A%20%222016-02-15T16%3A20%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22endhost/%20files%20seem%20alright%20to%20me%20other%20than%20what%20appears%20to%20be%20a%20typo%20%28see%20line%20note%29%22%2C%20%22created_at%22%3A%20%222016-02-15T17%3A59%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22Naming%20should%20now%20be%20consistent.%22%2C%20%22created_at%22%3A%20%222016-02-16T12%3A33%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22lgtm%21%20%28I%20didn%27t%20review%20topology/%2A%20%2C%20tests/%2A%20and%20tools/%2A%20as%20the%20PR%20is%20too%20big%20for%20github%3B-%29%5Cr%5CnNow%20we%20only%20need%20to%20change%20%60pcb%60%20to%20%60seg%60%20%3B-%29%22%2C%20%22created_at%22%3A%20%222016-02-16T15%3A39%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20685bac7f662fef94739e3748fd3a39117c49fb97%20endhost/sciond.py%20346%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53018578%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%60as_pairs%60%20here%20and%20below%22%2C%20%22created_at%22%3A%20%222016-02-16T14%3A41%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Good%20catch%2C%20fixed.%22%2C%20%22created_at%22%3A%20%222016-02-16T17%3A04%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/sciond.py%3AL394-441%22%7D%2C%20%22Pull%20a04b22056097a8fd5ee62f26050c41602d62052e%20endhost/scion_socket.py%20110%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r52927755%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22should%20be%20sa.isd_as%20%3D%20c_uint%28self.isd_as.int%28%29%29%22%2C%20%22created_at%22%3A%20%222016-02-15T17%3A56%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22Oops%2C%20fixed.%22%2C%20%22created_at%22%3A%20%222016-02-16T08%3A56%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_socket.py%3AL437-459%22%7D%2C%20%22Pull%20685bac7f662fef94739e3748fd3a39117c49fb97%20endhost/sciond.py%2079%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/630%23discussion_r53016669%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22btw.%20shouldn%27t%20%60update%28%29%60%20read%20first/last_ia%20itself%3F%22%2C%20%22created_at%22%3A%20%222016-02-16T14%3A27%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Quite%20possibly%2C%20but%20as%20this%20is%20quite%20old%2C%20i%27m%20going%20to%20leave%20it%20for%20now.%22%2C%20%22created_at%22%3A%20%222016-02-16T17%3A04%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/sciond.py%3AL156-176%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/630#issuecomment-184278734'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> @pszalach :)
  @aznair @ercanucan - can you please review the changes in endhost/, please? I haven't tested anything other than sciond in there.
- <a href='https://github.com/aznair'></a> endhost/ files seem alright to me other than what appears to be a typo (see line note)
- <a href='https://github.com/kormat'></a> Naming should now be consistent.
- <a href='https://github.com/pszalach'></a> lgtm! (I didn't review topology/\* , tests/\* and tools/\* as the PR is too big for github;-)
  Now we only need to change `pcb` to `seg` ;-)
- [ ] <a href='#crh-comment-Pull a04b22056097a8fd5ee62f26050c41602d62052e endhost/scion_socket.py 110'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/630#discussion_r52927755'>File: endhost/scion_socket.py:L437-459</a></b>
- <a href='https://github.com/aznair'></a> should be sa.isd_as = c_uint(self.isd_as.int())
- <a href='https://github.com/kormat'></a> Oops, fixed.
- [ ] <a href='#crh-comment-Pull 685bac7f662fef94739e3748fd3a39117c49fb97 endhost/sciond.py 79'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/630#discussion_r53016669'>File: endhost/sciond.py:L156-176</a></b>
- <a href='https://github.com/pszalach'></a> btw. shouldn't `update()` read first/last_ia itself?
- <a href='https://github.com/kormat'></a> Quite possibly, but as this is quite old, i'm going to leave it for now.
- [ ] <a href='#crh-comment-Pull 685bac7f662fef94739e3748fd3a39117c49fb97 endhost/sciond.py 346'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/630#discussion_r53018578'>File: endhost/sciond.py:L394-441</a></b>
- <a href='https://github.com/pszalach'></a> `as_pairs` here and below
- <a href='https://github.com/kormat'></a> Good catch, fixed.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/630?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/630?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/630'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: endhost/scion_socket.py [hunk 1]
```diff
@@ -437,24 +437,22 @@ class ScionClientSocket(ScionBaseSocket):
...
         """
         self.proto = proto
-        self.isd, self.ad = isd_ad
+        self.isd_as = isd_as
         self.target_IP, self.target_port = target_address
         self.libsock = CDLL(os.path.join(SHARED_LIB_LOCATION,
                                          SHARED_LIB_CLIENT))
         sa = C_SCIONAddr()
-        sa.isd_ad = c_uint(ISD_AD(self.isd, self.ad).int())
+        sa.isd_as = c_uint(self.isd_as).int()
```
Comment (aznair) [inline endhost/scion_socket.py]: should be sa.isd_as = c_uint(self.isd_as.int())

Comment (kormat) [inline endhost/scion_socket.py]: Oops, fixed.

File: endhost/sciond.py [hunk 2]
```diff
@@ -158,23 +156,20 @@ def handle_path_reply(self, pkt):
...
             self.requests.put((key, None))
 
-    def _handle_up_seg(self, pcb, first, last):
-        if self.addr.get_isd_ad() != (last.isd_id, last.ad_id):
+    def _handle_up_seg(self, pcb, first_ia, last_ia):
+        if self.addr.isd_as != last_ia:
             return
-        self.up_segments.update(pcb, first.isd_id, first.ad_id,
-                                last.isd_id, last.ad_id)
+        self.up_segments.update(pcb, first_ia, last_ia)
```
Comment (pszal) [inline endhost/sciond.py:162]: btw. shouldn't `update()` read first/last_ia itself?

File: endhost/sciond.py [hunk 3]
```diff
@@ -419,47 +394,47 @@ def _req_key_map(self, key, req_keys):
...
-            dst_core_ads.add(seg.get_first_pcbm().ad_id)
-        # Generate all possible AD pairs
-        ad_pairs = list(product(src_core_ads, dst_core_ads))
-        return self._find_core_segs(self.addr.isd_id, dst_isd, ad_pairs)
+            dst_core_ases.add(seg.get_first_pcbm().isd_as[1])
+        # Generate all possible AS pairs
+        as_pairs = list(product(src_core_ases, dst_core_ases))
+        return self._find_core_segs(self.addr.isd_as[0], dst_isd, as_pairs)
 
     def _find_core_segs(self, src_isd, dst_isd, ad_pairs):
```
Comment (pszal) [inline endhost/sciond.py]: `as_pairs` here and below

File: endhost/sciond.py (hunk 2 above)
Comment (kormat) [inline endhost/sciond.py:162]: Quite possibly, but as this is quite old, i'm going to leave it for now.

File: endhost/sciond.py (hunk 3 above)
Comment (kormat) [inline endhost/sciond.py]: Good catch, fixed.

**Pull request merged**